### ISSUE TITLE
gc: treat older manifest as unreachable?

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/Pkg.jl#2685]
I have manifests lying around from old Julia versions that I haven't used in ages but that are still there and which are therefore keeping old versions of packages and artifacts from being collected. When scanning `manifest_usage.toml`, I think it might make sense to ignore manifests that haven't been modified in some amount of time. Usage would be a better criterion, but we don't record when a manifest is used, only when it's modified. Perhaps that's close enough?

**Comment by KristofferC** [comment]
One possible scenario I see is that you have a long running julia service going and then you login to that computer and do a Pkg operation. Then automatic gc hits and remove all packages and artifacts used by the service.

**Comment by StefanKarpinski** [comment]
Good point. We could add the load path environments back into the root set even if they haven't been updated. After all, you definitely shouldn't delete anything that is currently loadable.

**Comment by IanButterworth** [comment]
Wouldn't that only guard against deleting/ignoring manifests in use in the current session, not others, like in the long-running service situation?

Perhaps the usage file needs both a latest used time, and a latest closed time, the latter being updated atexit or when switching environments, and items must be considered in the usage file if 
```
isnothing(last_closed_time) || last_used_time > last_closed_time
```
That might break backwards compatability for the usage file though

**Comment by DilumAluthge** [comment]
@maleadt any thoughts on this? I was thinking it could be useful for the issues we're having in https://github.com/JuliaCI/julia-buildkite-plugin/issues/17

**Comment by StefanKarpinski** [comment]
The issue with recording access time is that we don't want to put too much stuff in the code loading path. Even if we did, since code is typically only loaded early in a process, the scenario still exists where the long-running process is using something, logs that usage, then runs for a long time and some other Julia process is started and cleans up that other process's dependencies. The only thing that would prevent that entirely is to periodically re-record usage throughout the duration of the long-running process, which seems like an awful lot to do here.

**Comment by StefanKarpinski** [comment]
Ah, another possibility: record that a process started using a manifest and on exist that it stopped. Of course, if the program hard crashes or gets SIGKILLed then it won't write the stop and the "start usage" message becomes permanently unpaired. You could do some fancy stuff with recording the PID of the process and seeing if that process still exists, but yikes, that's getting complex.

**Comment by StefanKarpinski** [comment]
Maybe just make this opt-in with some flag to the `gc` command and maybe have some way to opt into that being the default. People generally know if they're running long-running processes or not.
